sys/arm64/vmm/vmm_hyp.c

```diff
@@ -0,0 +1,800 @@
+		default:
+		LOAD_APR(0);
+#undef LOAD_APR
+		}
+	}
+}
+
+static uint64_t
+vmm_hyp_call_guest(struct hyp *hyp, struct hypctx *hypctx)
+{
+	struct hypctx host_hypctx;
+	uint64_t cntvoff_el2;
+	uint64_t ich_hcr_el2, ich_vmcr_el2, cnthctl_el2, cntkctl_el1;
+	uint64_t ret;
+	uint64_t s1e1r, hpfar_el2;
+	bool hpfar_valid;
+
+	vmm_hyp_reg_store(&host_hypctx, NULL, false);
+
+	/* Save the host special registers */
+	cnthctl_el2 = READ_SPECIALREG(cnthctl_el2);
+	cntkctl_el1 = READ_SPECIALREG(cntkctl_el1);
+	cntvoff_el2 = READ_SPECIALREG(cntvoff_el2);
+
+	ich_hcr_el2 = READ_SPECIALREG(ich_hcr_el2);
+	ich_vmcr_el2 = READ_SPECIALREG(ich_vmcr_el2);
+
+	vmm_hyp_reg_restore(hypctx, hyp, true);
+
+	/* Load the common hypervisor registers */
+	WRITE_SPECIALREG(vttbr_el2, hyp->vttbr_el2);
+
+	host_hypctx.mdcr_el2 = READ_SPECIALREG(mdcr_el2);
+	WRITE_SPECIALREG(mdcr_el2, hypctx->mdcr_el2);
+
+	/* Call into the guest */
+	ret = vmm_enter_guest(hypctx);
+
+	WRITE_SPECIALREG(mdcr_el2, host_hypctx.mdcr_el2);
+	isb();
+
+	/* Store the exit info */
+	hypctx->exit_info.far_el2 = READ_SPECIALREG(far_el2);
+	vmm_hyp_reg_store(hypctx, hyp, true);
+
+	hpfar_valid = true;
+	if (ret == EXCP_TYPE_EL1_SYNC) {
+		switch(ESR_ELx_EXCEPTION(hypctx->tf.tf_esr)) {
+		case EXCP_INSN_ABORT_L:
+		case EXCP_DATA_ABORT_L:
+			/*
+			 * The hpfar_el2 register is valid for:
+			 *  - Translaation and Access faults.
+			 *  - Translaation, Access, and permission faults on
+			 *    the translation table walk on the stage 1 tables.
+			 *  - A stage 2 Address size fault.
+			 *
+			 * As we only need it in the first 2 cases we can just
+			 * exclude it on permission faults that are not from
+			 * the stage 1 table walk.
+			 *
+			 * TODO: Add a case for Arm erratum 834220.
+			 */
+			if ((hypctx->tf.tf_esr & ISS_DATA_S1PTW) != 0)
+				break;
+			switch(hypctx->tf.tf_esr & ISS_DATA_DFSC_MASK) {
+			case ISS_DATA_DFSC_PF_L1:
+			case ISS_DATA_DFSC_PF_L2:
+			case ISS_DATA_DFSC_PF_L3:
+				hpfar_valid = false;
+				break;
+			}
+			break;
+		}
+	}
+	if (hpfar_valid) {
+		hypctx->exit_info.hpfar_el2 = READ_SPECIALREG(hpfar_el2);
+	} else {
+		/*
+		 * TODO: There is a risk the at instruction could cause an
+		 * exception here. We should handle it & return a failure.
+		 */
+		s1e1r =
+		    arm64_address_translate_s1e1r(hypctx->exit_info.far_el2);
+		if (PAR_SUCCESS(s1e1r)) {
+			hpfar_el2 = (s1e1r & PAR_PA_MASK) >> PAR_PA_SHIFT;
+			hpfar_el2 <<= HPFAR_EL2_FIPA_SHIFT;
+			hypctx->exit_info.hpfar_el2 = hpfar_el2;
+		} else {
+			ret = EXCP_TYPE_REENTER;
+		}
+	}
+
+	vmm_hyp_reg_restore(&host_hypctx, NULL, false);
+
+	/* Restore the host special registers */
+	WRITE_SPECIALREG(ich_hcr_el2, ich_hcr_el2);
+	WRITE_SPECIALREG(ich_vmcr_el2, ich_vmcr_el2);
+
+	WRITE_SPECIALREG(cnthctl_el2, cnthctl_el2);
+	WRITE_SPECIALREG(cntkctl_el1, cntkctl_el1);
+	WRITE_SPECIALREG(cntvoff_el2, cntvoff_el2);
+
+	return (ret);
+}
+
+static uint64_t
+vmm_hyp_read_reg(uint64_t reg)
+{
+	switch(reg) {
```

Comment:
```suggestion
	switch (reg) {
```
Giving up commenting on all these; grep of the diff would be a more efficient way to find them all